core/src/stored/append.cc

```diff
@@ -52,39 +50,151 @@ static char OK_replicate[] = "3000 OK replicate data\n";
 
 void PossibleIncompleteJob(JobControlRecord* jcr, int32_t last_file_index) {}
 
+
+ProcessedFileData::ProcessedFileData(DeviceRecord* record)
+    : volsessionid_(record->VolSessionId)
+    , volsessiontime_(record->VolSessionTime)
+    , fileindex_(record->FileIndex)
+    , stream_(record->Stream)
+    , data_len_(record->data_len)
+    , data_(record->data, record->data + record->data_len)
+{
+}
+
+DeviceRecord ProcessedFileData::GetData()
+{
+  DeviceRecord devicerecord{};
+  devicerecord.VolSessionId = volsessionid_;
+  devicerecord.VolSessionTime = volsessiontime_;
+  devicerecord.FileIndex = fileindex_;
+  devicerecord.Stream = stream_;
+  devicerecord.data_len = data_len_;
+  devicerecord.data = data_.data();
+
+  return devicerecord;
+}
+
+ProcessedFile::ProcessedFile(int32_t fileindex) : fileindex_(fileindex) {}
+
+void ProcessedFile::SendAttributesToDirector(JobControlRecord* jcr)
+{
+  for_each(attributes_.begin(), attributes_.end(),
+           [&jcr](ProcessedFileData& attribute) {
+             DeviceRecord devicerecord = attribute.GetData();
+             SendAttrsToDir(jcr, &devicerecord);
+           });
+}
+
+void ProcessedFile::AddAttribute(DeviceRecord* record)
+{
+  attributes_.emplace_back(ProcessedFileData(record));
+}
+
+bool IsAttribute(DeviceRecord* record)
+{
+  return record->maskedStream == STREAM_UNIX_ATTRIBUTES
+         || record->maskedStream == STREAM_UNIX_ATTRIBUTES_EX
+         || record->maskedStream == STREAM_RESTORE_OBJECT
+         || CryptoDigestStreamType(record->maskedStream) != CRYPTO_DIGEST_NONE;
+}
+
+static void UpdateFileList(JobControlRecord* jcr)
+{
+  Dmsg0(100, _("... update file list\n"));
+  jcr->impl->dcr->DirAskToUpdateFileList(jcr);
+}
+
+static void UpdateJobmediaRecord(JobControlRecord* jcr)
+{
+  Dmsg0(100, _("... create job media record\n"));
+  jcr->impl->dcr->DirCreateJobmediaRecord(false);
+}
+
+static void UpdateJobrecord(JobControlRecord* jcr)
+{
+  Dmsg2(100, _("... update job record: %llu bytes %lu files\n"), jcr->JobBytes,
+        jcr->JobFiles);
+  jcr->impl->dcr->DirAskToUpdateJobRecord(jcr);
+}
+
+void DoBackupCheckpoint(JobControlRecord* jcr)
+{
+  Jmsg(jcr, M_INFO, 0,
+       _("Checkpoint: Syncing current backup status to catalog\n"));
+  UpdateJobrecord(jcr);
+  UpdateFileList(jcr);
+  UpdateJobmediaRecord(jcr);
+  Jmsg(jcr, M_INFO, 0, _("Checkpoint completed\n"));
+}
+
+static time_t DoTimedCheckpoint(JobControlRecord* jcr,
+                                time_t checkpoint_time,
+                                time_t checkpoint_interval)
+{
+  time_t now
+      = std::chrono::system_clock::to_time_t(std::chrono::system_clock::now());
+  if (now > checkpoint_time) {
+    checkpoint_time = now + checkpoint_interval;
```

Comment:
Is that what the user would expect?
If the checkpoint was delayed for some reason should the next checkpoint happen at
`now + checkpoint_interval`
or should it be more like
 `while(checkpoint_time <= now) checkpoint_time += checkpoint_interval;`

Reply:
I see where you're going, I agree, your solution keeps the pace of checkpoints more regular looking